### PR TITLE
Add block-by-id and id-by-height consensus endpoints

### PR DESCRIPTION
**Pull request by ChrisSchinnerl**
This PR adds 2 endpoints.
1. the `/consensus/blocks/:id` endpoint which allows a user to retrieve a whole block by id
2. the `/consensus/headers/:height` endpoint. This might return the whole header in the future but only returns the BlockID at the moment.

I chose 2 separate wildcards since one endpoint returns a block and the other one is mostly used to return information about the block. In the future we might also add `/consensus/blocks/:height` or `consensus/headers/:id` in case it becomes necessary.

### CODE REVIEW

**Comment by S-anasol** [comment]
Maybe you want add IDs(hashes) as keys then?
I think this is really useful. Since external apps cant calculate it on demand like it work in Sia wallet.
Using it for explorer.siahub for example.

https://github.com/NebulousLabs/Sia/pull/2046

**Comment by DavidVorick** [comment]
I believe this violates a standard API goal, which is that you should be able to fetch what you need in a single API call. If you want a block at a given height, you currently need to make 2 queries, one to get the header, and then another to get the block resulting from that header.

Some people want to be able to get the raw block. Some want to be able to get the json block. Some want the header, and some want the id. I think we might be able to return all 4 things (potentially a fifth if we return both the raw header and the json header) each block. It means that a query has a lot of data, but typically if you are making a query for a block you are expecting a lot of data.

I'm guessing @lukechampine has some thoughts.

**Comment by ChrisSchinnerl** [comment]
I think it is not unreasonable to have 1 wildcard for blocks and 1 for headers. That way you can get a whole block by id and height and also just the header by id and height. At the moment the header endpoint only returns the id and the `consensus/blocks/:height` and `consensus/headers/:id` endpoints are missing, but that can be extended/added on demand.

**Comment by lukechampine** [comment]
@DavidVorick and I discussed this offline and we agree that a single endpoint is probably the best approach. For now, it will support only two query params:
```
/consensus/blocks?id=foo
/consensus/blocks?height=1
```
Later, we can add a `&raw=true` param for returning base64-encoded binary blocks, and perhaps `startheight` and `endheight` params for returning a range of blocks. (That's why it's `blocks` and not `block`.) But for now, just those two are fine.

**Comment by magmel48** [comment]
is there any chance to have this PR rewritten (in terms of documentation only) and merged?

File: types/encoding.go [hunk 1]
```diff
@@ -280,6 +280,11 @@ func (bid BlockID) String() string {
 	return fmt.Sprintf("%x", bid[:])
 }
 
+// LoadString loads a BlockID from a string
+func (bid *BlockID) LoadString(str string) error {
+	return (*crypto.Hash)(bid).LoadString(str)
+}
```
Comment (lukechampine) [inline types/encoding.go]: :ok_hand:

File: node/api/client/consensus.go [hunk 2]
```diff
@@ -1,9 +1,26 @@
...
+
+// ConsensusBlocksIDGet requests the /consensus/blocks api resource
+func (c *Client) ConsensusBlocksIDGet(id types.BlockID) (block types.Block, err error) {
+	err = c.Get("/consensus/blocks?id="+id.String(), &block)
+	return
+}
+
+// ConsensusBlocksHeightGet requests the /consensus/blocks api resource
+func (c *Client) ConsensusBlocksHeightGet(height types.BlockHeight) (block types.Block, err error) {
+	err = c.Get("/consensus/blocks?height="+strconv.FormatUint(uint64(height), 10), &block)
```
[the remark this answers is not in the document]
Comment (lukechampine) [reply node/api/client/consensus.go]: we may even want to add a `Getf` method

**Pull request merged**